### ISSUE TITLE
SFTP using ssh key

### ISSUE THREAD

**Issue by bryanberger** [ccrisan/motioneye#1007]
Add an option to upload via SFTP using pub/priv keys rather than username/password.

Could you possibly do this by mounting `rw` and adding a `.ssh` config and keypair?

possible here?
https://github.com/ccrisan/motioneye/blob/master/motioneye/uploadservices.py#L803

**Comment by ccrisan** [comment]
@bryanberger I agree, this feature makes sense. Pull requests are welcomed, by the way.

**Comment by eBeagle** [comment]
Sorry for the bump. Any progress on this feature?

**Comment by TheFlatEric** [comment]
So, I have good news and bad news about using pub/priv keys with SFTP.

The **tl;dr** version: Good news, I forked the repo and got a version of this working.  Bad news, there seem to be some bugs in the underlying libraries that I haven't solved that make me wary of trying to make a pull request.  As a result, these changes are not really production ready.  If you want to see what I have done, you are welcome to look at the dev-sftp-key branch over here: https://github.com/TheFlatEric/motioneye/tree/dev-sftp-key  Please note if you want to use them you will need to do some command line work to setup the keys.

Longer technical details:

I've been playing with this software on raspberry pis, so most of my focus was on getting it working in motioneyeos.  I forked both repos and created branches from master called "dev-sftp-key".  I had been using the last release version, so I only wanted to change from that release and not worry about ongoing dev changes.  I have locally compiled builds running on 2 different raspberry pis that I've been using for testing.

The current state of the changes is that I added a "Keyfile" input box to the web interface that only appears when SFTP Server is selected.  It is expecting the location of the private keyfile on the OS, and it will assume the public keyfile is in the same location and the file name is the same with an additional ".pub" extension.  If a value is entered in the Keyfile input, then the code will attempt to use the username and key info while ignoring the password.

The problems come in due to the way that libcurl is being built.  libcurl uses libssh2, which on many debian based distributions uses libgcrypt.  This results in some differences with how OpenSSH works that have lead to two issues that I have encountered so far.

1. libgcrypt does not like the keyfiles created by OpenSSH's ssh-keygen tool by default.  I found out more about that [here](https://stackoverflow.com/questions/14760798/curl-sftp-public-key-authentication-fails-callback-error).  Fortunately, there are flags that can be passed to ssh-keygen to force generation in the PEM format.  You can find out more about that [here](https://serverfault.com/questions/706336/how-to-get-a-pem-file-from-ssh-key-pair).  I ended up re-coding the key using this command "ssh-keygen -p -m PEM -f id_rsa -e > id_rsa.pem"
2. Even after fixing the keyfiles, the remote server is logging errors every time a connection occurs.  The logs are full of:
```
Dec 31 11:47:52 sshd: Bad packet length 2700599005.
Dec 31 11:47:52 sshd: Fssh_ssh_dispatch_run_fatal: Connection from user x xx.xx.xx.xx port 39382: message authentication code incorrect
```
What I read suggests this is some sort of mismatch in the encryption, but I have not been able to figure out what to do about it.  Maybe upgrading libssh2 would help, but I'm not sure.  This seems to happen after the file is uploaded, so it isn't a total blocker, but I don't like filling up server logs with errors.

Because of the second issue, I paused doing any further work.  If people consider the issues trivial, then the UI end of my changes likely could use some polish to make it so that the user does not have use the command line to create and use keys.

Alternatively, it may be better to just create a shell script to use the OpenSSH tools (ssh/sftp/scp) and call that shell script using the "Run A Command" option.  I am considering going that route as a test next.

All that being said, I hope what I found helps.  A Happy New Year to you all.
